feat: added async loading support to medusa

SGFGOV opened the pull request:
added support to async load modules fixes issue #140

SGFGOV commented:
@adrien2p  can you please review this

adrien2p commented:
Thank you for the contribution. I ll have a look as soon as I can 🚀

adrien2p commented:
have you tested it? cause medusa will still strip the properties from the config no?

adrien2p commented on src/Medusa.ts:
```diff
@@ -45,11 +46,30 @@ export class Medusa {
+			configModule: ConfigModule;
+			configFilePath: string;
+		};
+		const resolveConfigProperties = async (obj): Promise<ConfigModule> => {
```
todo: should be moved to utilities and if you can add a new unit test for it please

SGFGOV replied:
moved to utilities, need to add a unit test

SGFGOV replied:
added unit test

adrien2p commented on src/Medusa.ts:
```diff
@@ -45,11 +46,30 @@ export class Medusa {
+			monitoring: MonitoringOptions;
+		};
+
+		/*const { configModule } = getConfigFile(this.#rootDir, 'medusa-config') as {
```
todo: rm

SGFGOV replied:
done

adrien2p commented on src/cli/commands/migrate.ts:
```diff
@@ -32,7 +42,23 @@ type Options = { run: boolean; revert: boolean; show: boolean; tenant_codes: str
+		configModule: ConfigModule;
+		configFilePath: string;
+	};
+	const resolveConfigProperties = async (obj): Promise<ConfigModule> => {
```
todo: from the comment above you will be able to use the util instead of duplicating the code

SGFGOV replied:
done

adrien2p commented on src/cli/commands/migrate.ts:
```diff
@@ -32,7 +42,23 @@ type Options = { run: boolean; revert: boolean; show: boolean; tenant_codes: str
  */
 export async function migrate({ run, revert, show, tenant_codes }: Options): Promise<void> {
-	const { configModule } = getConfigFile(process.cwd(), `medusa-config`) as { configModule: ConfigModule };
+	//const { configModule } = getConfigFile(process.cwd(), `medusa-config`) as { configModule: ConfigModule };
```
todo: rm

SGFGOV replied:
done

adrien2p commented on src/cli/commands/migrate.ts:
```diff
@@ -1,15 +1,25 @@
-import { createConnection } from 'typeorm';
+import { createConnection, LoggerOptions } from 'typeorm';
```
suggestion: Would be nice to look into medusa and apply the same concept for the feature flag, so that with the extender migration command you can also run the medusa migration and that the feature flags are taken into account and also usable for custom migrations. wdyt?

SGFGOV replied:
well in the sequence medusa migrations need to run before medusa extender migrations, the feature flag is automatically handled isn't it ?

SGFGOV replied:
@adrien2p  please let me know if anything else needs to be done

SGFGOV commented:
yes.. the properties you need have to be inside the project config.

Best Regards
Govind

On Thu, 20 Oct 2022, 18:33 Adrien de Peretti, ***@***.***>
wrote:

> have you tested it? cause medusa will still strip the properties from the
> config no?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/adrien2p/medusa-extender/pull/139#issuecomment-1285505024>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXEQJHHNCHYMCYEKZUYLSFDWEE7KVANCNFSM6AAAAAARIAIYCM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

-- 

This message (including any attachments) may contain confidential, 
proprietary, privileged and/or private
information. The information is 
intended to be for the use of the individual or entity designated above. If
you are not the intended recipient of this message, please notify the 
sender immediately, and delete the
message and any attachments. Any 
disclosure, reproduction, distribution or other use of this message or
any 
attachments by an individual or entity other than the intended recipient is 
prohibited.   

adrien2p commented:
Will you be able to tackle the comments?

SGFGOV commented:
> have you tested it? cause medusa will still strip the properties from the config no?

yup, you need to encapsulate settings inside project config

adrien2p commented:
Could you add everything related to vscode in the gitignore as well as add unit tests for the async load with config that also includes sync/async function to resolve?

SGFGOV commented:
> Could you add everything related to vscode in the gitignore as well as add unit tests for the async load with config that also includes sync/async function to resolve?

done

adrien2p commented:
> > Could you add everything related to vscode in the gitignore as well as add unit tests for the async load with config that also includes sync/async function to resolve?
> 
> 
> 
> done

Ahah thanks, but it is still not exactly how it should be done. The tests for the utility should be in a separate file as it concerns the util and not the cli command it self. 

A bit more structure in the tests would be nice.

The fake config data can be moved to a fixtures directory in the tests dir like it is done in other places. You can then create different config files for the different use case and just load them without having to writeRile each time and without having hard readable dummy config 🙏

Can you clean up please?

After that we should be good or I will directly write the proposed changes so that you can either do it or accept it depending on the type of changes.

Thanks a lot again 👌

SGFGOV commented:
> Owner

Thanks a lot Adrien :), I'll be obliged if you can help with the clean up.

SGFGOV commented:
@adrien2p  i've refactored as requested

adrien2p commented:
> @adrien2p  i've refactored as requested

Ahah this is not that yet 😂 sorry.
I ll write you the changes asap

adrien2p commented on src/cli/tests/cli.spec.ts:
```diff
@@ -1,6 +1,9 @@
 import * as path from 'path';
 import { exec, ExecException } from 'child_process';
 import { normalizeString } from './__utils__/normalizeString';
+import { asyncLoadConfig } from '../utils/async-load-config';
```
are those import used?

adrien2p commented on src/cli/tests/cli.spec.ts:
```diff
@@ -10,7 +13,7 @@ function cli(args, cwd): Promise<{ code: number; error: ExecException; stdout: s
 				resolve({
 					code: error && error.code ? error.code : 0,
-					error,
+					error: error as ExecException,
```
is that necessary?

SGFGOV replied:
yes, else it throws this error 
type 'ExecException | null' is not assignable to type 'ExecException'.
  Type 'null' is not assignable to type 'ExecException'.ts(2322)

adrien2p replied:
In that case you shouldn't cast it to this type but instead update the return type to include null

adrien2p commented on src/cli/utils/async-load-config.ts:
```diff
@@ -0,0 +1,22 @@
+import { ConfigModule } from 'core/types';
+import { getConfigFile } from 'medusa-core-utils';
+
+export async function asyncLoadConfig(): Promise<ConfigModule> {
```
would be nice to pass the path as the argument of the function `configPath: string`

SGFGOV replied:
we could do that, but by default medusa expects the configfile path to be in the root folder, I didn't want to change that behaviour

adrien2p replied:
It would just give more flexibility and would make it explicit, the same way you have to pass the path to medusa config loader. Wdyt?

adrien2p commented on src/cli/utils/fixtures/asyn-load-config-test-fixture.ts:
```diff
@@ -0,0 +1,67 @@
+export const asyncDataWithAsyncFunctionAndAsyncParameter = `		const axios=require('axios').default
```
create one fixture file for each test config and pass the path to that file to the util method, also I dont think that axios is used no?

In the end, each fixture file is like a medusa-config-file
medusa-config-sync.js
medusa-config-async.js
medusa-config-sync-async.js

SGFGOV replied:
it will depend on making the path configurable, as medusa's default behaviour is to only pick up medusa-config.js, I didn't want to alter that. 
wdyt ?

adrien2p replied:
In any case you can do it as it would just mean that the test need to cp the file in a different Location and the rm the file but that is fine

SGFGOV replied:
done

adrien2p commented on src/cli/utils/tests/async-load-config.spec.ts:
```diff
@@ -0,0 +1,45 @@
+
+		const pathToConfigFile = `${process.cwd()}/../medusa-config.js`;
+		writeFileSync(pathToConfigFile, data);
+		const configModule = await asyncLoadConfig();
```
don't need to create the file as the fixture will be consumed, instead just pass the path to the fixture file config to test

SGFGOV replied:
it will depend on making the path configurable, as medusa's default behaviour is to only pick up medusa-config.js, I didn't want to alter that.
wdyt ?

SGFGOV replied:
dones

adrien2p commented on src/cli/utils/tests/async-load-config.spec.ts:
```diff
@@ -0,0 +1,45 @@
+		writeFileSync(pathToConfigFile, data);
+		const configModule = await asyncLoadConfig();
+		expect(configModule).toBeDefined();
+		expect(configModule.projectConfig).toBeDefined;
```
you must check that each props have the expected value, otherwise we can't be sure it works properly just with the test

adrien2p replied:
same for the other tests

adrien2p commented on src/Medusa.ts:
```diff
@@ -45,12 +46,9 @@ export class Medusa {
-			configModule: {
-				monitoring: MonitoringOptions;
-			};
+		const configModule = (await asyncLoadConfig()) as unknown as {
```
Question: would it be nice to make the utils template so that the return type is specified instead of force cast the type?

SGFGOV replied:
done

adrien2p commented on src/cli/commands/migrate.ts:
```diff
@@ -31,8 +15,10 @@ type Options = { run: boolean; revert: boolean; show: boolean; tenant_codes: str
+
 export async function migrate({ run, revert, show, tenant_codes }: Options): Promise<void> {
-	const { configModule } = getConfigFile(process.cwd(), `medusa-config`) as { configModule: ConfigModule };
+	//const { configModule } = getConfigFile(process.cwd(), `medusa-config`) as { configModule: ConfigModule };
```
Rm

SGFGOV replied:
done

adrien2p commented:
I think that after resolving the last few comments we should be good to merge

adrien2p commented:
@SGFGOV let me know when you have time to finish it 😀

adrien2p commented:
@SGFGOV should I consider that pr as dead?

SGFGOV commented:
> @SGFGOV should I consider that pr as dead?

Hi Sorry, I've just been held up the last two weeks. No its not dead, I'll finish it in a day or two. Sorry again

adrien2p commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "medusa-extender",
-  "version": "1.7.7",
+  "version": "1.7.6",
```
There is an issue here, it should not be changed

SGFGOV replied:
oops my bad sorry

adrien2p commented on src/cli/tests/cli.spec.ts:
```diff
@@ -97,4 +98,28 @@ describe('CLI', () => {
 		);
 	});
+
+	it('should aysnc load medusa-config', async () => {
```
Should be removed from here

adrien2p commented on src/cli/utils/fixtures/asyncDataWithAsyncFunctionAndAsyncParameter.js:
```diff
@@ -0,0 +1,27 @@
+			return Promise.resolve("test");
+		}
+
+		async function asyncConfig(){
```
This is not necessary, you can just return the config in the export

SGFGOV replied:
i don't understand this

adrien2p replied:
Why don't you just export the config instead of having an intermediary method? We don't need to have a method that return the config and then a method that await it and then export the last method.
We can just have

async function password()...
module.exports = { ... }

SGFGOV replied:
done

adrien2p commented on src/cli/utils/fixtures/asyncFunctionWithAsyncData.js:
```diff
@@ -0,0 +1,22 @@
+async function asyncConfig(){
```
This is not necessary, you can just return the config in the export

adrien2p replied:
@SGFGOV let me know when you have the time to finalize :) can't wait to merge that

SGFGOV replied:
sorry,  will try to complete it tomorrow.

adrien2p commented on src/cli/utils/tests/async-load-config.spec.ts:
```diff
@@ -0,0 +1,43 @@
+import { asyncLoadConfig } from '../async-load-config';
+
+const pathToConfigFile = `${process.cwd()}/medusa-config.js`;
+const runLoadTest = async (testFixtureFileName: string) => {
```
You can rename the method to getConfig and use it in the test case, each test case should be able to explicitly check the values of each params of the fake config and not just instance of promise etc since they should be resolved already

SGFGOV replied:
this line does it expect(configModule.projectConfig).toMatchObject(testProjectConfig);

adrien2p reviewed (changes requested):
Thanks for the changes, but still need a bit more changes. I think like it lacks of rigor when it comes to write tests, I let you look at my comment. Otherwise nice work man 🔥

adrien2p commented:
Thanks for your work 🚀 could I get you to run the linter please

SGFGOV commented:
I think the github action permissions needed to be changed as we were writing a file in the earlier implementation, I've modified it so that optionally you can pass a file to the loader (as you had suggested right in the beginning :-) . I didn't understand why you said that then. Now I do 👍 ) and also modified the tests. Now the tests are also in line with the expected behaviour.

adrien2p commented:
Looks much better 🥰 happy that you finally see what i was aiming for 🚀 nice work man have you been able to lint as well?

adrien2p commented:
@SGFGOV ? I d like to merge it this week and release a new version

SGFGOV commented:
sure

SGFGOV commented:
yes I did lint

adrien2p commented on src/cli/utils/fixtures/async-function-with-async-parameter.js:
```diff
@@ -0,0 +1,27 @@
+}
+
+  
+  module.exports = asyncConfig();
```
todo: missing new line

adrien2p commented on src/cli/utils/fixtures/async-function-with-non-async-data.js:
```diff
@@ -0,0 +1,21 @@
+async function asyncConfig(){
+
```
todo: rm

adrien2p commented on src/cli/utils/fixtures/async-function-with-non-async-data.js:
```diff
@@ -0,0 +1,21 @@
+      },
+      plugins:[],
+    }
+
```
todo: rm

adrien2p commented on src/cli/utils/fixtures/async-function-with-non-async-data.js:
```diff
@@ -0,0 +1,21 @@
+}
+
+  
+  module.exports = asyncConfig();
```
todo: missing new line

adrien2p commented on src/cli/utils/fixtures/async-parameter.js:
```diff
@@ -0,0 +1,18 @@
+			database_password:password
+		  },
+		  plugins:[],
+		};
```
todo: indent issue + todo: missing new line

adrien2p reviewed (changes requested):
LGTM, few comment and I merge

adrien2p commented:
@SGFGOV Let me know when you have the time for the lasts clean up stuff so that I merge it :)

adrien2p commented:
@SGFGOV any news?

SGFGOV commented:
my apologies..have been under the weather the past couple of days

adrien2p commented:
Looks like something went wrong with your merge

SGFGOV commented:
yeah .. looks like something went grossly wrong, let me close and reopen the pr on a fresh branch
